security/pkg/nodeagent/sds/sds_stream_test.go

```diff
@@ -0,0 +1,394 @@
+	}
+	// Send first request and verify response
+	if err := stream.Send(req); err != nil {
+		notifyChan <- notifyMsg{Err: err, Message: fmt.Sprintf("stream: stream.Send failed: %v", err)}
+	}
+	_, err := stream.Recv()
+	if err != nil {
+		notifyChan <- notifyMsg{Err: err, Message: BlockGenSecretError}
+	}
+	notifyChan <- notifyMsg{Err: nil, Message: "notify push secret 2"}
+}
+
+type StreamSetup struct {
+	t                          *testing.T
+	socket                     string
+	server                     *Server
+	secretStore                *mockIngressGatewaySecretStore
+	initialTotalPush           float64
+	initialTotalUpdateFailures float64
+}
+
+func (s *StreamSetup) waitForSDSReady() error {
+	var conErr, streamErr error
+	var conn *grpc.ClientConn
+	for i := 0; i < 20; i++ {
+		if conn, conErr = setupConnection(s.socket); conErr == nil {
+			sdsClient := sds.NewSecretDiscoveryServiceClient(conn)
+			header := metadata.Pairs(credentialTokenHeaderKey, fakeToken1)
+			ctx := metadata.NewOutgoingContext(context.Background(), header)
+			if _, streamErr = sdsClient.StreamSecrets(ctx); streamErr == nil {
+				return nil
+			}
+		}
+		time.Sleep(10 * time.Millisecond)
+	}
+	return fmt.Errorf("cannot connect SDS server, connErr: %v, streamErr: %v", conErr, streamErr)
+}
+
+type mockIngressGatewaySecretStore struct {
+	checkToken      bool
+	secrets         sync.Map
+	mutex           sync.RWMutex
+	secretCacheHit  int
+	secretCacheMiss int
+}
+
+func (ms *mockIngressGatewaySecretStore) SecretCacheHit() int {
+	ms.mutex.RLock()
+	defer ms.mutex.RUnlock()
+	return ms.secretCacheHit
+}
+
+func (ms *mockIngressGatewaySecretStore) SecretCacheMiss() int {
+	ms.mutex.RLock()
+	defer ms.mutex.RUnlock()
+	return ms.secretCacheMiss
+}
+
+func (ms *mockIngressGatewaySecretStore) GenerateSecret(ctx context.Context, conID, resourceName, token string) (*model.SecretItem, error) {
+	if ms.checkToken && token != fakeToken1 && token != fakeToken2 {
+		return nil, fmt.Errorf("unexpected token %q", token)
+	}
+
+	key := cache.ConnKey{
+		ConnectionID: conID,
+		ResourceName: resourceName,
+	}
+	if resourceName == testResourceName {
+		s := &model.SecretItem{
+			CertificateChain: fakeCertificateChain,
+			PrivateKey:       fakePrivateKey,
+			ResourceName:     testResourceName,
+			Version:          time.Now().Format("01-02 15:04:05.000"),
+			Token:            token,
+		}
+		fmt.Println("Store secret for key: ", key, ". token: ", token)
+		ms.secrets.Store(key, s)
+		// if it is the invalid connection for test,
+		// we still store the secret to the cache, however we will throw an error here
+		// to earlier terminate the flow and then test whether the secret will be deleted
+		// after connection is stopped
+
+		if strings.Contains(conID, "invalid") {
+			return s, fmt.Errorf("invalid connection for test")
+		}
+		return s, nil
+	}
+
+	if resourceName == cache.RootCertReqResourceName || strings.HasPrefix(resourceName, "file-root:") {
```

Comment:
I think this block is never used by the test. Shall we remove it?

Reply:
ok